x/concentrated-liquidity/internal/math/math.go

```diff
@@ -94,23 +94,52 @@ func CalcAmount1Delta(liq, sqrtPriceA, sqrtPriceB sdk.Dec, roundUp bool) sdk.Dec
 	return liq.Mul(diff)
 }
 
-// getNextSqrtPriceFromAmount0RoundingUp utilizes the current squareRootPrice, liquidity of denom0, and amount of denom0 that still needs
-// to be swapped in order to determine the next squareRootPrice
-// Note: we are using only using the precise formula here.
-func GetNextSqrtPriceFromAmount0RoundingUp(sqrtPriceCurrent, liquidity, amountRemaining sdk.Dec) (sqrtPriceNext sdk.Dec) {
-	if amountRemaining.Equal(sdk.ZeroDec()) {
+// GetNextSqrtPriceFromAmount0InRoundingUp utilizes sqrtPriceCurrent, liquidity, and amount of denom0 that still needs
+// to be swapped in order to determine the sqrtPriceNext.
+// When we swap for token one out given token zero in, the price is decreasing, and we need to move the sqrt price (decrease it) less
+// to avoid overpaying the amount out of the pool. Therefore, we round up.
+// sqrt_next = liq * sqrt_cur / (liq + token_in * sqrt_cur)
+func GetNextSqrtPriceFromAmount0InRoundingUp(sqrtPriceCurrent, liquidity, amountZeroRemainingIn sdk.Dec) (sqrtPriceNext sdk.Dec) {
+	if amountZeroRemainingIn.Equal(sdk.ZeroDec()) {
 		return sqrtPriceCurrent
 	}
 
-	product := amountRemaining.Mul(sqrtPriceCurrent)
+	product := amountZeroRemainingIn.Mul(sqrtPriceCurrent)
 	denominator := liquidity.Add(product)
 	return liquidity.Mul(sqrtPriceCurrent).QuoRoundUp(denominator)
 }
 
-// getNextSqrtPriceFromAmount1RoundingDown utilizes the current squareRootPrice, liquidity of denom1, and amount of denom1 that still needs
-// to be swapped in order to determine the next squareRootPrice
-func GetNextSqrtPriceFromAmount1RoundingDown(sqrtPriceCurrent, liquidity, amountRemaining sdk.Dec) (sqrtPriceNext sdk.Dec) {
-	return sqrtPriceCurrent.Add(amountRemaining.QuoTruncate(liquidity))
+// GetNextSqrtPriceFromAmount0OutRoundingUp utilizes sqrtPriceCurrent, liquidity, and amount of denom0 that still needs
+// to be swapped out order to determine the sqrtPriceNext.
+// When we swap for token one in given token zero out, the price is increasing and we need to move the price up enough
+// so that we get the desired output amount out. Therefore, we round up.
+// sqrt_next = liq * sqrt_cur / (liq - token_out * sqrt_cur)
+func GetNextSqrtPriceFromAmount0OutRoundingUp(sqrtPriceCurrent, liquidity, amountZeroRemainingOut sdk.Dec) (sqrtPriceNext sdk.Dec) {
+	if amountZeroRemainingOut.Equal(sdk.ZeroDec()) {
+		return sqrtPriceCurrent
+	}
+
+	product := amountZeroRemainingOut.Mul(sqrtPriceCurrent)
+	denominator := liquidity.Sub(product)
```

Comment:
Note: the difference is in subtraction here.

While there could be ways to merge the two methods, I find it extremely helpful for readability and clarity to have separate APIs depending on which token is being swapped in and in which direction.

Reply:
yeah also a +1 in leaving them separately for low math equations